src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -645,26 +645,29 @@ private Uri GetContainingEntitySetUri(Uri baseUri, ODataUri odataUri)
         {
             ODataPath path = odataUri.Path;
             List<ODataPathSegment> segments = path.ToList();
+            int lastIndex = segments.Count - 1;
             ODataPathSegment lastSegment = segments.Last();
             while (!(lastSegment is NavigationPropertySegment) && !(lastSegment is OperationSegment))
             {
-                segments.Remove(lastSegment);
-                lastSegment = segments.Last();
+                lastIndex--;
+                lastSegment = segments[lastIndex];
             }
 
             // also removed the last navigation property segment
-            segments.Remove(lastSegment);
+            segments = segments.GetRange(0, lastIndex);
 
             // Remove the unnecessary type cast segment.
-            ODataPathSegment nextToLastSegment = segments.Last();
-            while (nextToLastSegment is TypeSegment)
+            lastIndex = segments.Count - 1;
+            lastSegment = segments[lastIndex];
+            while (lastSegment is TypeSegment)
             {
-                ODataPathSegment previousSegment = segments[segments.Count - 2];
+                ODataPathSegment previousSegment = segments[lastIndex - 1];
                 IEdmStructuredType owningType = previousSegment.TargetEdmType as IEdmStructuredType;
                 if (owningType != null && owningType.FindProperty(lastSegment.Identifier) != null)
                 {
-                    segments.Remove(nextToLastSegment);
-                    nextToLastSegment = segments.Last();
+                    segments.RemoveAt(lastIndex);
+                    --lastIndex;

```

Comment:
@marabooy Any particular reason why you chose to use `--lastIndex` here but `lastIndex--` [here](https://github.com/OData/odata.net/pull/1854/files#diff-5c1c8c475ffa3bfbcec9695a7554881aR652)?